File: C++/Tree/144_Binary Tree Preorder Traversal.cpp
/////////////////// Recursive /////////////////

class Solution {
public:
    vector<int> ans;
    vector<int> preorderTraversal(TreeNode* root) {
        if(root != NULL){
            ans.push_back(root->val);
            preorderTraversal(root->left);
            preorderTraversal(root->right);
        }
        return ans;
    }
};

/////////////////// iteravtive ///////////////////

class Solution {
public:

    vector<int> preorderTraversal(TreeNode* root) {
      vector<int> ans;
      stack<TreeNode*> st;
        st.push(root);
        
        while(!st.empty()){
            TreeNode* temp = st.top();
            st.pop();
            if(temp != nullptr){
                ans.push_back(temp->val);
                st.push(temp->right);
                st.push(temp->left);
            }
            
        }
        
        return ans;
    }
};


///////////////////////////// Moris Traversal //////////////////////////

class Solution {
public:

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> preorder;
        
        TreeNode* curr = root;
        while(curr != nullptr){
            if(curr->left == nullptr){
                preorder.push_back(curr->val);
                curr = curr->right;
            }
            else{
                TreeNode* prev = curr->left;
                while(prev->right && prev->right != curr){
                    prev = prev->right;
                }
                
                if(prev->right == nullptr){
                    prev ->right = curr;
                    preorder.push_back(curr->val);
                    curr = curr->left;
                }
                else{
                    prev->right = nullptr;
                    curr = curr->right;
                }
    
            }
 
        }
        return preorder;
          
    }
};

File: Python/Tree/144_Binary Tree Preorder Traversal.py
////////////// iterative  ////////////////////////

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        
        stack = [root]
        
        while stack:
            curr_node = stack.pop()
            if curr_node:
                ans.append(curr_node.val)
                stack.append(curr_node.right)
                stack.append(curr_node.left)
        return ans

//////////////////  Recursive //////////////

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        res = []
        def dfs(root):
            if not root:
                return
            res.append(root.val)
            
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        return res


////////////////// Moris Traversal ////////////////

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        preorder = []
        
        curr = root
        
        while curr:
            if(curr.left == None):
                preorder.append(curr.val)
                curr = curr.right
            else:
                last = curr.left
                
                while last.right and last.right != curr:
                    last = last.right
                    
                    
                if(last.right == None):
                    preorder.append(curr.val)
                    last.right = curr
                    curr = curr.left
                else:
                    last.right = None
                    curr = curr.right
        return preorder
                    
